### PR TITLE
Add top-nσ sampler

### PR DESCRIPTION
**Pull request by Lazarus-931**
Adds top-nσ sampling (arXiv 2411.07641): keeps tokens with logit ≥ max − n·σ over the raw logits, invariant to temperature. Exposed via `make_sampler`, `generate_step`, and the server API as `top_n_sigma` (default 0 = off). Deterministic tests included. Part of Blaizzy/mlx-vlm#1648.

### CODE REVIEW

File: mlx_vlm/sample_utils.py [hunk 1]
```diff
@@ -151,6 +158,39 @@ def apply_top_k(
...
+            values are in the ~0.5-2.0 range.
+    """
+    if n_sigma < 0:
+        raise ValueError(
+            f"`top_n_sigma` has to be a non-negative float, but is {n_sigma}"
+        )
+    return _top_n_sigma(logits, n_sigma)
+
+
+@partial(mx.compile, inputs=mx.random.state, outputs=mx.random.state)
```
Comment (lucasnewman) [inline mlx_vlm/sample_utils.py]: This doesn't use categorical sampling / random state, do we really need to capture it?

**Review by lucasnewman — APPROVED** [review]
See one comment but looks good modulo the merge conflict.

File: mlx_vlm/sample_utils.py (hunk 1 above)
Comment (Lazarus-931) [reply mlx_vlm/sample_utils.py]: hm, yes I'll remove that, did so to keep repo convention

**Review by lucasnewman — APPROVED** [review]

**Comment by Blaizzy** [comment]
Please resolve the conflicts before we merge :)

**Comment by Lazarus-931** [comment]
good to merge @Blaizzy

**Review by Blaizzy — APPROVED** [review]
LGTM, thanks!
